File: src/db-adapter/repository/sql/product.go
package sql

import (
	"context"

	pb "github.com/dynatrace/easytrade/dbadapter/proto"
	"github.com/dynatrace/easytrade/dbadapter/repository"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Product struct {
	Id       *uuid.UUID `gorm:"primaryKey;default:(-)"`
	Name     string
	Ppt      float64
	Currency string
}

func (Product) TableName() string { return repository.TableProducts }

var _ repository.ProductRepository = (*ProductRepository)(nil)

type ProductRepository struct{ db *gorm.DB }

func NewProductRepository(db *gorm.DB) repository.ProductRepository {
	return &ProductRepository{db: db}
}

func (repo *ProductRepository) GetAll(ctx context.Context) ([]*pb.ProductMessage, error) {
	return findAndMapAll(repo.db.WithContext(ctx), (*Product).toProto)
}

func (src *Product) toProto() *pb.ProductMessage {
	return &pb.ProductMessage{
		Id:       uuidString(src.Id),
		Name:     src.Name,
		Ppt:      src.Ppt,
		Currency: src.Currency,
	}
}

File: src/broker-service/src/Entities/Products/Product.cs
namespace EasyTrade.BrokerService.Entities.Products;

public class Product(string name, decimal ppt, string currency)
{
    public Guid Id { get; set; }
    public string Name { get; set; } = name;
    public decimal Ppt { get; set; } = ppt;
    public string Currency { get; set; } = currency;

    public Product(Guid id, string name, decimal ppt, string currency)
        : this(name, ppt, currency)
    {
        Id = id;
    }
}
